sql/sql_admission_control.h

```diff
@@ -0,0 +1,356 @@
+ * make a decision of whether or not a resource can be allocated. Connections
+ * and query limits are among the resources multi-tenancy plugin allocates.
+ * Note that admission control is now part of the multi-tenancy module to
+ * hanlde query throttling.
+ *
+ * The isolation level is defined by an entity. The entity could be a database,
+ * a user, or any thing that multi-tenancy plugin defines to isolate the
+ * resource allocation.
+ *
+ * See sql_admission_control.cc for implementation.
+ */
+
+extern int multi_tenancy_admit_query(THD *, const MT_RESOURCE_ATTRS *);
+extern int multi_tenancy_exit_query(THD *, const MT_RESOURCE_ATTRS *);
+
+/**
+  Per-thread information used in admission control.
+*/
+
+#ifdef HAVE_PSI_INTERFACE
+
+static PSI_mutex_key key_ac_node_lock;
+static PSI_cond_key key_ac_node_cond;
+
+static PSI_mutex_info all_ac_node_mutexes[] = {
+    {&key_ac_node_lock, "st_ac_node::lock", PSI_FLAG_SINGLETON, 0,
+     PSI_DOCUMENT_ME}};
+
+static PSI_cond_info all_ac_node_conds[] = {
+    {&key_ac_node_cond, "st_ac_node::cond", PSI_FLAG_SINGLETON, 0,
+     PSI_DOCUMENT_ME}};
+
+static void init_st_ac_node_keys() {
+  const char *const category = "sql";
+  int count;
+
+  count = static_cast<int>(array_elements(all_ac_node_mutexes));
+  mysql_mutex_register(category, all_ac_node_mutexes, count);
+
+  count = static_cast<int>(array_elements(all_ac_node_conds));
+  mysql_cond_register(category, all_ac_node_conds, count);
+}
+#endif
+
+struct st_ac_node {
+  mysql_mutex_t lock;
+  mysql_cond_t cond;
+  st_ac_node() {
+#ifdef HAVE_PSI_INTERFACE
+    init_st_ac_node_keys();
+#endif
+    mysql_mutex_init(key_ac_node_lock, &lock, MY_MUTEX_INIT_FAST);
+    mysql_cond_init(key_ac_node_cond, &cond);
+  }
+
+  ~st_ac_node() {
+    mysql_mutex_destroy(&lock);
+    mysql_cond_destroy(&cond);
+  }
+};
+
+/**
+  Class used in admission control.
+
+  Every entity (database or table or user name) will have this
+  object created and stored in the global map AC::ac_map.
+
+*/
+class Ac_info {
+  friend class AC;
+#ifdef HAVE_PSI_INTERFACE
+  PSI_mutex_key key_lock;
+  PSI_mutex_info key_lock_info[1] = {{&key_lock, "Ac_info::lock", 0}};
+#endif
+  // Queue to track the running and waiting threads.
+  std::deque<std::shared_ptr<st_ac_node>> queue;
+  std::atomic<unsigned long> waiting_queries;
+  // Protects the queue.
+  mysql_mutex_t lock;
+
+ public:
+  Ac_info() {
+#ifdef HAVE_PSI_INTERFACE
+    mysql_mutex_register("sql", key_lock_info, array_elements(key_lock_info));
+#endif
+    mysql_mutex_init(key_lock, &lock, MY_MUTEX_INIT_FAST);
+    waiting_queries = 0;
+  }
+  ~Ac_info() { mysql_mutex_destroy(&lock); }
+  // Disable copy constructor.
+  Ac_info(const Ac_info &) = delete;
+  Ac_info &operator=(const Ac_info &) = delete;
+};
+
+/**
+  Global class used to enforce per admission control limits.
+*/
+class AC {
+  // This map is protected by the rwlock LOCK_ac.
+  std::unordered_map<std::string, std::shared_ptr<Ac_info>> ac_map;
```

Comment:
usings/typedefs for custom containers
```suggestion
  using Ac_info_ptr_container = std::unordered_map<std::string, Ac_info_ptr;
  Ac_info_ptr_container ac_map;
```